From c43f1abb087f90d728ace458df99f10d55442254 Mon Sep 17 00:00:00 2001
From: Rezvan Mahdavi Hezaveh <rezvan@chromium.org>
Date: Fri, 14 Feb 2025 22:27:19 +0000
Subject: [PATCH] [explicit-resource-management] Call dispose method on correct
 receiver

This CL fixes the bug of calling dispose method on undefined receiver.

Bug: 396661134
Change-Id: I0121318ddc3afe61867ca34625061ba9a67db9b9
Reviewed-on: https://chromium-review.googlesource.com/c/v8/v8/+/6271683
Reviewed-by: Shu-yu Guo <syg@chromium.org>
Commit-Queue: Rezvan Mahdavi Hezaveh <rezvan@chromium.org>
Cr-Commit-Position: refs/heads/main@{#98712}
---
 .../builtins-async-disposable-stack.cc        |  5 +--
 ...anagement-call-sync-from-async-receiver.js | 34 +++++++++++++++++++
 2 files changed, 37 insertions(+), 2 deletions(-)
 create mode 100644 test/mjsunit/harmony/explicit-resource-management-call-sync-from-async-receiver.js

diff --git a/src/builtins/builtins-async-disposable-stack.cc b/src/builtins/builtins-async-disposable-stack.cc
index 777e6551eb3..a1f2b557a25 100644
--- a/src/builtins/builtins-async-disposable-stack.cc
+++ b/src/builtins/builtins-async-disposable-stack.cc
@@ -77,6 +77,7 @@ BUILTIN(AsyncDisposeFromSyncDispose) {
   //      captures method and performs the following steps when called:
   //        a. Let O be the this value.
   //        b. Let promiseCapability be ! NewPromiseCapability(%Promise%).
+  DirectHandle<Object> receiver = args.receiver();
   DirectHandle<JSPromise> promise = isolate->factory()->NewJSPromise();
 
   //        c. Let result be Completion(Call(method, O)).
@@ -90,8 +91,8 @@ BUILTIN(AsyncDisposeFromSyncDispose) {
   try_catch.SetVerbose(false);
   try_catch.SetCaptureMessage(false);
 
-  MaybeDirectHandle<Object> result = Execution::Call(
-      isolate, sync_method, isolate->factory()->undefined_value(), {});
+  MaybeDirectHandle<Object> result =
+      Execution::Call(isolate, sync_method, receiver, {});
 
   DirectHandle<Object> result_handle;
 
diff --git a/test/mjsunit/harmony/explicit-resource-management-call-sync-from-async-receiver.js b/test/mjsunit/harmony/explicit-resource-management-call-sync-from-async-receiver.js
new file mode 100644
index 00000000000..d31cf5a7ac9
--- /dev/null
+++ b/test/mjsunit/harmony/explicit-resource-management-call-sync-from-async-receiver.js
@@ -0,0 +1,34 @@
+// Copyright 2025 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Flags: --js-staging
+
+let value = false;
+
+const obj =
+    {
+      [Symbol.dispose]() {
+        value = (this === obj)
+      }
+    }
+
+async function TestAwaitUsing() {
+  await using x = obj;
+}
+
+async function TestAsyncDisposableStack() {
+  const stack = new AsyncDisposableStack();
+  stack.use(obj);
+  stack[Symbol.asyncDispose]();
+}
+
+async function RunTest() {
+  await TestAwaitUsing();
+  assertSame(true, value);
+  value = false;
+  await TestAsyncDisposableStack();
+  assertSame(true, value);
+}
+
+RunTest();
-- 
2.39.5 (Apple Git-154)

